src/Wrapper.js

```diff
@@ -117,30 +117,57 @@ class Wrapper extends React.Component {
       direct: path.length === 1,
       intent: transaction && this.makeTransactionIntent(bag),
       paths: path.length ? [path] : [],
+      pretransaction: (transaction && transaction.pretransaction) || null,
     }
   }
   handleTransaction = bag => {
-    const { intent, direct, paths } = this.reshapeTransactionBag(bag)
-    this.showWeb3ActionSigner(intent, { direct, error: null, paths })
-  }
-  handleSigningWeb3Tx = (transaction, intent) => {
     const {
-      walletWeb3,
-      transactionBag: { accept, reject },
-    } = this.props
+      intent,
+      direct,
+      paths,
+      pretransaction,
+    } = this.reshapeTransactionBag(bag)
+    this.showWeb3ActionSigner(intent, {
+      direct,
+      error: null,
+      paths,
+      pretransaction,
+    })
+  }
 
-    walletWeb3.eth.sendTransaction(transaction, (err, res) => {
-      this.handleSignerClose()
+  async signWeb3Tx(transaction, intent) {
+    const { walletWeb3 } = this.props
+    return new Promise((resolve, reject) => {
+      walletWeb3.eth.sendTransaction(transaction, (err, res) => {
+        if (err) {
+          reject(err)
+        } else {
+          resolve(res)
+        }
+      })
+    })
+  }
+
+  handleSigningWeb3Tx = async (transaction, intent, pretransaction) => {
+    const { transactionBag } = this.props
 
-      if (err) {
-        this.showWeb3ActionSigner(intent, { error: err })
-        console.error(err)
-        reject(err)
-        return
+    try {
+      if (pretransaction) {
+        await this.signWeb3Tx(pretransaction, intent)
       }
 
-      accept(res)
-    })
+      const transactionRes = await this.signWeb3Tx(transaction, intent)
+
+      transactionBag.accept(transactionRes)
+      this.handleSignerClose()
+
+      // Display an error in the panel if a transaction fail
+    } catch (err) {
+      console.error(err)
```

Comment:
Not sure how easy this would be to do with your signing panel changes @bpierre, but if were easy with the restructuring, it would be nice to show a different error if the pretransaction failed.